### ISSUE TITLE
imageObj is undefined when shape is something else than 'image"

### ISSUE THREAD

**Issue by Herment** [almende/vis#3532]
Hi there,

I've found an issue: 

When I do new
`vis.Network(canvasContainer, graphData, options);`
and options has
`"nodes":{"shape":"ellipse"}`
then even if all nodes of graphData has images, when I do
`option = {
"nodes":{
"shape":"ellipse"
}
};
graphCanvas.setOptions(option);`
I get "TypeError:this.imageObj is undefined".

This issue is really a blocker for me, and I think it can be for others to.

Thanks a lot! Hope you'll fix it soon!

**Comment by wimrijnders** [comment]
There are several things that can go wrong here. I'm afraid that you will have to be more specific.

How did you define your image nodes? Are the images defined in the nodes or in the options? Please give an example network definition that illustrates the problem.

**Comment by Herment** [comment]
Ok so : 
I have a script that create me nodes, like that

```
{
	"id":"9120",
	"hidden":false,
	"label":"MyNode",
	"color":{
		"background":"#acd7e5",
		"border":"#acd7e5",
		"highlight":{
			"background":"#acd7e5",
			"border":"#343434"
		},
		"hover":{
			"background":"#acd7e5",
			"border":"#acd7e5"
		}
	},
	"cluster_id":"",
	"image":"style/topology/img/router_ERROR.svg",
	"x":110,
	"y":322
}
```
and I put it in a JSON array.
After that, I use a function in my JavaScript to create the canvas:

```
function createGraphNodes(GraphTemp) {
	GraphTemp = JSON.parse(GraphTemp);
	var nodes = new vis.DataSet(GraphTemp["NDA"]); // The list of all my nodes
	var edges = new vis.DataSet(GraphTemp["LDA"]); // The list of all links between nodes
	graphData = {
		nodes:nodes,
		edges:edges,
		parameters:GraphTemp["PDA"]
	};
}
```

The graphData variable is used in 

`vis.Network(canvasContainer, graphData, options);`

I hope this will help you. I'll be pleased to answer other questions.

Thanks

**Comment by wimrijnders** [comment]
There should be a `shape:'image'` somewhere, either in the nodes themselves or in the options[I]. Where is it in your case?

I don't see it in your example, so I assume it's in the options. Is this correct?

----
[I] in a group is also possible.

**Comment by Herment** [comment]
Yes sure!

My options are:

o```
ptions = {
	"edges":{
		"smooth":false,
		"arrows":{
			"to":{
				"enabled":false
			}
		}
	},
	"physics":{
		"enabled":false
	},
	"interaction":{
		"navigationButtons":true,
		"hover":true,
		"multiselect":true,
		"selectable":true
	},
	"layout":{
		"hierarchical":{
			"enabled":false,
			"direction":"UD"
		}
	},
	"nodes":{
		"borderWidthSelected": 3,
		"shape":"image",
	    "size": 20
	}
};
```

**Comment by wimrijnders** [comment]
OK. It just dawned on me that is is similar to #3438. I will check if the solution there is appropriate for this.

But I'll follow the procedure for now: first confirm.

**Comment by Herment** [comment]
Hum... I think that the problem is not the same that #3438. There is no relation with edges, physic or spinning problem. It's only on images.

Can you explain me why you think it's similar? Maybe I just don't see the relation.

**Comment by wimrijnders** [comment]
Sorry, typo. Editing it. Should be #3483.

**Comment by Herment** [comment]
Sadly, I tested the fix and it doesn't work. 
That's what I updated in my options:

```
"layout":{
	"hierarchical":{
		"enabled":false,
		"direction":"UD"
	},
   "improvedLayout": false // The update
}
```

**Comment by wimrijnders** [comment]
Well, you can't take over the workaround verbatim. #3483 is a special case of this one. The solution would need to be more generalized.

**Comment by Herment** [comment]
I see. So I hope that the problem could be fix as soon as possible.
Thanks a lot for your advises and your work on this project!

**Comment by wimrijnders** [comment]
Yeah, sure. I have time availability issues, so please hang on. Will do my best.

**Comment by wimrijnders** [comment]
Sorry can't reproduce. I made a test page based on the data you gave me, but it appears to be working as expected.

[This is the demo](https://jsfiddle.net/wimrijnders/k98vutg5/) - can you tell me if anything extra needs to be added?

**Comment by Herment** [comment]
Can you try this instead?

https://jsfiddle.net/k98vutg5/2/

The version of vis.js is 4.20.1

**Comment by wimrijnders** [comment]
Ah right! It's the other way around, going from `ellipse` to `image` shape. I'm on it.

**Comment by wimrijnders** [comment]
OK. With the added knowledge, I found the problem within two minutes[I].

Any option field `image` is ignored when initializing nodes, if the shape is not `image` (or `circularImage`) to begin with. Thus, when `image` is actually specified in the options, the href to the image is just not present.

In the current initialization setup, there is no way to retrieve the image automatically.

**Workaround:**

Assuming there's a reason that the images are specified in the node instances, setting the shape `image` in the node instances directly will work:

```js
	var nodesData = data.nodes.get();
	for (var n = 0; n < nodesData.length; ++n) {
		nodesData[n].shape = 'image';
	}
	data.nodes.update(nodesData);
```
This comes in place of the call to `network.setOptions()`.

**Thoughts on Fix:**

From the point of view of the `Network` code, this is actually the intended behavior. However, due to [recent discussion](https://github.com/almende/vis/pull/3330#issuecomment-332659181), I have come to realize that:

- the `DataSet` contents are leading for the user
- having discrepancies in the user data and the network state leads to confusion
- therefore, the `DataSet` contents and the `Network` state should be effectively synchronized at all times.

So, rather than being a purist about this, I'll go the extra length and read out any `image` fields present from the nodes `DataSet` when the shape changes[II]. I hope you don't mind if I put this as low-priority; it's too late to get it in the next release and the workaround should be good for you.

---
[I] Note to self: ask for demo's sooner and more often.
[II] It should be obvious I'm opening Pandora's box here; if this should be done for the `image` field, what about the rest???

**Comment by wimrijnders** [comment]
Upon glaring at the code, I see a much easier way to make the fix: just preload any images encountered in the settings and use them when you need them. The overhead is acceptable, unless you have a lot of distinct images and many, many nodes.

Wil go with this.

**Comment by Herment** [comment]
Ok, I will use it for the moment. Thanks a lot for your help!

**Comment by wimrijnders** [comment]
My pleasure! Until next time!
